securesystemslib/dsse.py

```diff
@@ -0,0 +1,159 @@
+    ):
+        self.payload = payload
+        self.payload_type = payload_type
+        self.signatures = signatures
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Envelope):
+            return False
+
+        return (
+            self.payload == other.payload
+            and self.payload_type == other.payload_type
+            and self.signatures == other.signatures
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "Envelope":
+        """Creates a DSSE Envelope from its JSON/dict representation.
+
+        Arguments:
+            data: A dict containing a valid payload, payloadType and signatures
+
+        Raises:
+            KeyError: If any of the "payload", "payloadType" and "signatures"
+                fields are missing from the "data".
+
+            FormatError: If signature in "signatures" is incorrect.
+
+        Returns:
+            A "Envelope" instance.
+        """
+
+        payload = b64dec(data["payload"])
+        payload_type = data["payloadType"]
+
+        signatures = [
+            Signature.from_dict(signature) for signature in data["signatures"]
+        ]
+
+        return cls(payload, payload_type, signatures)
+
+    def to_dict(self) -> dict:
+        """Returns the JSON-serializable dictionary representation of self."""
+
+        return {
+            "payload": b64enc(self.payload),
+            "payloadType": self.payload_type,
+            "signatures": [
+                signature.to_dict() for signature in self.signatures
+            ],
+        }
+
+    def pae(self) -> bytes:
+        """Pre-Auth-Encoding byte sequence of self."""
+
+        return b"DSSEv1 %d %b %d %b" % (
+            len(self.payload_type),
+            self.payload_type.encode("utf-8"),
+            len(self.payload),
+            self.payload,
+        )
+
+    def sign(self, signer: Signer) -> Signature:
+        """Sign the payload and create the signature.
+
+        Arguments:
+            signer: A "Signer" class instance.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        signature = signer.sign(self.pae())
+        self.signatures.append(signature)
+
+        return signature
+
+    def verify(self, keys: List[Key], threshold: int) -> Dict[str, Key]:
+        """Verify the payload with the provided Keys.
+
+        Arguments:
+            keys: A list of public keys to verify the signatures.
+            threshold: Number of signatures needed to pass the verification.
+
+        Raises:
+            ValueError: If "threshold" is not valid.
+            VerificationError: If the enclosed signatures do not pass the
+                verification.
+
+        Note:
+            Mandating keyid in signatures and matching them with keyid of Key
+            in order to consider them for verification, is not DSSE spec
+            compliant (Issue #416).
+
+        Returns:
+            accepted_keys: A dict of unique public keys.
```

Comment:
Minor nit: Sphinx renders `accepted_keys` as return type, which is incorrect. Besides, the name is an implementation detail of this method, so no need to mention it in the docstring.

Reply:
Also, @jku made a good point about the return value being too vaguely defined, and not part of the dsse spec ( see https://github.com/secure-systems-lab/securesystemslib/pull/487#discussion_r1092886518)

I suggest we either return `None` -- I checked in-toto, we don't use it there -- or we add a tiny bit more detail in the documentation to explains that these are only the first ``threshold`` of ``keys`` that verified signatures.

Reply:
ping: https://github.com/secure-systems-lab/securesystemslib/pull/487/files#r1113994810

Reply:
Let's just add the docs here and move on. What do you think about...
```suggestion
            A dict of the threshold of unique public keys that verified a signature.
```
?